File: src-tauri/src/commands.rs
//! Tauri command handlers for the worldbuilding application.
//!
//! These commands bridge the frontend (Svelte/JavaScript) and backend (Rust) functionality.
//! All commands are async-capable and automatically manage thread safety via Tauri's State system.

use crate::licensing;
use crate::licensing::License;
use crate::models::{BrokenLink, FullPageData, PageHeader};
use crate::{
    config,
    error::Result,
    fonts, importer,
    models::{FileNode, RenderedPage},
    template,
    world::World,
};
use chrono::{Local, NaiveDate};
use std::path::PathBuf;
use tauri::{command, AppHandle, State};
use tauri_plugin_opener::OpenerExt;
use tracing::instrument;

// --- Vault and Initialization ---

/// Retrieves the stored vault path from the configuration file.
#[command]
#[instrument(skip(app_handle))]
pub fn get_vault_path(app_handle: AppHandle) -> Result<Option<String>> {
    config::get_vault_path(&app_handle)
}

/// Sets the vault path, saves it to config, and initializes the world state.
/// This uses fine-grained locking internally instead of a single write lock on the world.
#[command]
#[instrument(skip(world, app_handle))]
pub fn initialize_vault(path: String, world: State<World>, app_handle: AppHandle) -> Result<()> {
    world.change_vault(path, app_handle)
}

// --- Data Retrieval ---

/// Returns the tag index, mapping tags to lists of pages that contain them.
#[command]
#[instrument(skip(world))]
pub fn get_all_tags(world: State<World>) -> Result<Vec<(String, Vec<PageHeader>)>> {
    world.get_all_tags()
}

/// Returns the hierarchical file tree structure of the vault.
#[command]
#[instrument(skip(world))]
pub fn get_file_tree(world: State<World>) -> Result<FileNode> {
    world.get_file_tree()
}

/// Returns a list of all directory paths in the vault.
#[command]
#[instrument(skip(world))]
pub fn get_all_directory_paths(world: State<World>) -> Result<Vec<PathBuf>> {
    world.get_all_directory_paths()
}

/// Returns a list of all broken links in the vault.
#[command]
#[instrument(skip(world))]
pub fn get_all_broken_links(world: State<World>) -> Result<Vec<BrokenLink>> {
    world.get_all_broken_links()
}

// --- Page Rendering and Content ---

/// Processes raw markdown content, renders it to HTML with wikilinks resolved,
/// and returns a structured object for the frontend preview.
#[command]
#[instrument(skip(content, world))]
pub fn render_page_preview(content: String, world: State<World>) -> Result<RenderedPage> {
    world.render_page_preview(&content)
}

/// Parses the file on disk, renders the markdown to HTML, and returns a composed
/// object containing the raw content, and the rendered preview.
#[command]
#[instrument(skip(world))]
pub fn build_page_view(path: String, world: State<World>) -> Result<FullPageData> {
    world.build_page_view(&path)
}

/// Renders a string of pure Markdown to a `RenderedPage` object containing only HTML.
/// This command does not process wikilinks or frontmatter.
#[command]
#[instrument(skip(content, world))]
pub fn render_markdown(content: String, world: State<World>) -> Result<RenderedPage> {
    world.render_markdown(&content)
}

/// Converts a relative or absolute image path to a Base64 Data URL string.
#[command]
#[instrument(skip(world))]
pub fn get_image_as_base64(path: String, world: State<World>) -> Result<String> {
    world.get_image_as_base64(&path)
}

// --- File and Folder Operations ---

/// Writes content to a page on disk. The file watcher will pick up the change.
#[command]
#[instrument(skip(world, content))]
pub fn write_page_content(world: State<World>, path: String, content: String) -> Result<()> {
    world.write_page_content(&path, &content)
}

/// Creates a new, empty markdown file and synchronously updates the index.
#[command]
#[instrument(skip(world))]
pub fn create_new_file(
    world: State<World>,
    parent_dir: String,
    file_name: String,
    template_path: Option<String>,
) -> Result<PageHeader> {
    world.create_new_file(parent_dir, file_name, template_path)
}

/// Creates a new, empty folder.
#[command]
#[instrument(skip(world))]
pub fn create_new_folder(
    world: State<World>,
    parent_dir: String,
    folder_name: String,
) -> Result<()> {
    world.create_new_folder(parent_dir, folder_name)
}

/// Renames a file or folder on disk (in-place) and updates the index.
#[command]
#[instrument(skip(world))]
pub fn rename_path(world: State<World>, path: String, new_name: String) -> Result<()> {
    world.rename_path(PathBuf::from(path), new_name)
}

/// Deletes a file or folder from disk and updates the index.
#[command]
#[instrument(skip(world))]
pub fn delete_path(world: State<World>, path: String) -> Result<()> {
    world.delete_path(PathBuf::from(path))
}

/// Moves a file or folder to a new directory and updates the index and backlinks.
#[command]
#[instrument(skip(world))]
pub fn move_path(world: State<World>, source_path: String, dest_dir: String) -> Result<()> {
    world.move_path(PathBuf::from(source_path), PathBuf::from(dest_dir))
}

/// Duplicates a page, creating a new file with a numerical suffix.
#[command]
#[instrument(skip(world))]
pub fn duplicate_page(path: String, world: State<World>) -> Result<PageHeader> {
    world.duplicate_page(path)
}

/// Opens the specified path in the OS's default file explorer.
#[command]
pub fn open_in_explorer(app_handle: AppHandle, path: String) -> Result<()> {
    app_handle.opener().open_path(path, None::<&str>)?;
    Ok(())
}

// --- Importer ---

/// Imports a list of .docx files, converting them to Markdown.
#[command]
#[instrument(skip(world, app_handle))]
pub fn import_docx_files(
    world: State<World>,
    app_handle: AppHandle,
    docx_paths: Vec<PathBuf>,
) -> Result<Vec<PathBuf>> {
    world.import_docx_files(&app_handle, docx_paths)
}

/// Scans a directory for .docx files and imports them.
#[command]
#[instrument(skip(world, app_handle))]
pub fn import_docx_from_folder(
    world: State<World>,
    app_handle: AppHandle,
    folder_path: PathBuf,
) -> Result<Vec<PathBuf>> {
    world.import_docx_from_folder(&app_handle, folder_path)
}

/// Imports a MediaWiki XML dump file.
#[command]
#[instrument(skip(world))]
pub async fn import_mediawiki_dump(
    world: State<'_, World>,
    xml_path: PathBuf,
) -> Result<Vec<PathBuf>> {
    world.import_mediawiki_dump(xml_path).await
}

/// Checks if Pandoc is installed in the application's config directory.
#[command]
#[instrument(skip(app_handle))]
pub fn is_pandoc_installed(app_handle: AppHandle) -> Result<bool> {
    importer::is_pandoc_installed(&app_handle)
}

/// Downloads and extracts Pandoc to the application's config directory.
#[command]
#[instrument(skip(app_handle))]
pub async fn download_pandoc(app_handle: AppHandle) -> Result<()> {
    importer::download_pandoc(app_handle).await
}

// --- Licensing ---

/// Retrieves the current license status from the stored license file.
#[command]
#[instrument(skip(app_handle))]
pub fn get_license_status(app_handle: AppHandle) -> Result<Option<License>> {
    licensing::load_license(&app_handle)
}

/// Verifies a license key, and if valid, saves it to the config directory.
#[command]
#[instrument(skip(app_handle, license_key))]
pub async fn verify_and_store_license(
    app_handle: AppHandle,
    license_key: String,
) -> Result<License> {
    let license = licensing::validate_license(&license_key).await?;
    licensing::save_license(&app_handle, &license)?;
    Ok(license)
}

// --- System ---

/// Checks for the "APPIMAGE" environment variable to determine if the
/// application is running as an AppImage on Linux.
#[command]
#[instrument]
pub fn get_linux_install_type() -> String {
    // The APPIMAGE env var is set by the AppImage runtime.
    if std::env::var("APPIMAGE").is_ok() {
        "appimage".to_string()
    } else {
        // This indicates it's likely a .deb, .rpm, or other system-managed package.
        "other".to_string()
    }
}

/// Checks the number of days the application has been in use.
/// If it's the first time this check is run, it records the current date.
#[command]
#[instrument(skip(app_handle))]
pub fn get_app_usage_days(app_handle: AppHandle) -> Result<i64> {
    let mut config = config::load(&app_handle)?;

    match config.first_launch_date {
        Some(date_str) => {
            // If a date is already stored, calculate the difference.
            let first_launch_date = NaiveDate::parse_from_str(&date_str, "%Y-%m-%d")
                .unwrap_or_else(|_| Local::now().date_naive());
            let current_date = Local::now().date_naive();
            let duration = current_date.signed_duration_since(first_launch_date);
            Ok(duration.num_days())
        }
        None => {
            // If no date is stored, this is the first launch.
            // Record today's date and return 0 days.
            let today = Local::now().date_naive().format("%Y-%m-%d").to_string();
            config.first_launch_date = Some(today);
            config::save(&app_handle, &config)?;
            Ok(0)
        }
    }
}

// --- Template Commands ---

/// Retrieves a list of all available templates.
#[command]
#[instrument(skip(app_handle))]
pub fn list_templates(app_handle: AppHandle) -> Result<Vec<PageHeader>> {
    template::list_templates(&app_handle)
}

/// Reads the raw content of a specific template file.
#[command]
#[instrument]
pub fn read_template(path: String) -> Result<String> {
    template::read_template(&PathBuf::from(path))
}

/// Saves content to a template file.
#[command]
#[instrument(skip(app_handle, content))]
pub fn write_template(app_handle: AppHandle, name: String, content: String) -> Result<PathBuf> {
    template::write_template(&app_handle, &name, &content)
}

/// Deletes a template file.
#[command]
#[instrument]
pub fn delete_template(path: String) -> Result<()> {
    template::delete_template(&PathBuf::from(path))
}

// --- Custom Fonts ---

/// Scans the application's config directory for user-provided font files.
#[command]
#[instrument(skip(app_handle))]
pub fn get_user_fonts(app_handle: AppHandle) -> Result<Vec<fonts::UserFont>> {
    fonts::get_user_fonts(&app_handle)
}

File: src/lib/commands.ts
/**
 * @file This file serves as the API bridge between the Svelte frontend and the Rust backend.
 * It provides strongly-typed wrappers around Tauri's `invoke` function for every command
 * defined in the backend, ensuring type safety and a single point of entry for all
 * backend communication.
 */

import { invoke } from "@tauri-apps/api/core";
import type {
    FileNode,
    FullPageData,
    License,
    PageHeader,
    RenderedPage,
    TagMap,
    BrokenLink,
    UserFont,
} from "./bindings";

// --- Vault Commands ---

/**
 * Retrieves the stored vault path from the configuration file.
 * @returns A promise that resolves to the vault path string or null if not set.
 */
export const getVaultPath = () => invoke<string | null>("get_vault_path");

/**
 * Sets the vault path, saves it to config, and initializes the world state.
 * @param path The absolute path to the new vault directory.
 * @returns A promise that resolves when the vault is successfully initialized.
 */
export const initializeVault = (path: string) =>
    invoke<void>("initialize_vault", { path });

// --- Sidebar and Indexing Commands ---

/**
 * Returns the hierarchical file tree structure of the vault.
 * @returns A promise that resolves to the root FileNode of the vault.
 */
export const getFileTree = () => invoke<FileNode>("get_file_tree");

/**
 * Returns the tag index mapping tags to lists of pages that contain them.
 * @returns A promise that resolves to a map of tags to page paths.
 */
export const getAllTags = () => invoke<TagMap>("get_all_tags");

/**
 * Returns a list of all directory paths in the vault.
 * @returns A promise that resolves to an array of directory path strings.
 */
export const getAllDirectoryPaths = () =>
    invoke<string[]>("get_all_directory_paths");

/**
 * Returns a list of all broken links in the vault.
 * @returns A promise that resolves to an array of BrokenLink objects.
 */
export const getAllBrokenLinks = () =>
    invoke<BrokenLink[]>("get_all_broken_links");

// --- Page & File Operation Commands ---

/**
 * Parses a file on disk, renders it, and returns all data needed for the file view.
 * @param path The path to the file to build the view for.
 * @returns A promise that resolves to the complete data for the page view.
 */
export const buildPageView = (path: string) =>
    invoke<FullPageData>("build_page_view", { path });

/**
 * Writes new content to a page on disk.
 * @param path The path of the file to write to.
 * @param content The new markdown content to save.
 * @returns A promise that resolves when the file has been written.
 */
export const writePageContent = (path: string, content: string) =>
    invoke("write_page_content", { path, content });

/**
 * Renders a preview of markdown content without saving it to disk.
 * @param content The raw markdown content to render.
 * @returns A promise that resolves to the rendered page data.
 */
export const renderPagePreview = (content: string) =>
    invoke<RenderedPage>("render_page_preview", { content });

/**
 * Renders pure markdown content (no wikilink resolution, or YAML frontmatter)
 * @param content The raw markdown content to render.
 * @returns A promise that resolves to the rendered page data.
 */
export const renderMarkdown = (content: string) =>
    invoke<RenderedPage>("render_markdown", { content });

/**
 * Creates a new, empty markdown file.
 * @param parentDir The directory where the new file should be created.
 * @param fileName The name for the new file.
 * @param templatePath Optional path to a template file to use.
 * @returns A promise that resolves to the header data of the newly created page.
 */
export const createNewFile = (
    parentDir: string,
    fileName: string,
    templatePath?: string | null,
) =>
    invoke<PageHeader>("create_new_file", {
        parentDir,
        fileName,
        templatePath,
    });

/**
 * Creates a new, empty folder.
 * @param parentDir The directory where the new folder should be created.
 * @param folderName The name for the new folder.
 */
export const createNewFolder = (parentDir: string, folderName: string) =>
    invoke<void>("create_new_folder", { parentDir, folderName });

/**
 * Renames a file or folder in-place.
 * @param path The current path of the item to rename.
 * @param newName The new name for the item.
 */
export const renamePath = (path: string, newName: string) =>
    invoke<void>("rename_path", { path, newName });

/**
 * Deletes a file or folder.
 * @param path The path of the item to delete.
 */
export const deletePath = (path: string) =>
    invoke<void>("delete_path", { path });

/**
 * Moves a file or folder to a new directory.
 * This command delegates path construction to the backend, making it platform-safe.
 * @param sourcePath The full path of the item to move.
 * @param destDir The full path of the target directory.
 * @returns A promise that resolves when the item has been moved.
 */
export const movePath = (sourcePath: string, destDir: string) =>
    invoke<void>("move_path", { sourcePath, destDir });

/**
 * Duplicates a page, creating a new file with a numerical suffix.
 * @param path The path of the file to duplicate.
 * @returns A promise that resolves to the header data of the newly created page.
 */
export const duplicatePage = (path: string) =>
    invoke<PageHeader>("duplicate_page", { path });

/**
 * Opens a given path in the OS's default file explorer.
 * @param path The absolute path to the directory or file to open.
 */
export const openInExplorer = (path: string) =>
    invoke<void>("open_in_explorer", { path });

/**
 * Converts an image at a given path to a Base64 Data URL.
 * @param path The absolute or relative path to the image file.
 * @returns A promise that resolves to the Base64 Data URL string.
 */
export const getImageAsBase64 = (path: string) =>
    invoke<string>("get_image_as_base64", { path });

// --- Importer Commands ---

/**
 * Checks if Pandoc is installed in the application's config directory.
 * @returns A promise that resolves to true if Pandoc is found, false otherwise.
 */
export const isPandocInstalled = () => invoke<boolean>("is_pandoc_installed");

/**
 * Downloads and extracts Pandoc to the application's config directory.
 * @returns A promise that resolves when Pandoc has been successfully downloaded.
 */
export const downloadPandoc = () => invoke<void>("download_pandoc");

/**
 * Imports a list of .docx files, converting them to Markdown.
 * @param docxPaths An array of paths to the .docx files to import.
 * @returns A promise that resolves to an array of paths of the newly created Markdown files.
 */
export const importDocxFiles = (docxPaths: string[]) =>
    invoke<string[]>("import_docx_files", { docxPaths });

/**
 * Imports all .docx files found within a given folder.
 * @param folderPath The absolute path to the folder to scan.
 * @returns A promise that resolves to an array of paths of the newly created Markdown files.
 */
export const importDocxFromFolder = (folderPath: string) =>
    invoke<string[]>("import_docx_from_folder", { folderPath });

/**
 * Imports a MediaWiki XML dump file.
 * @param xmlPath The path to the MediaWiki XML dump file.
 * @returns A promise that resolves to an array of paths of the newly created Markdown files.
 */
export const importMediawikiDump = (xmlPath: string) =>
    invoke<string[]>("import_mediawiki_dump", { xmlPath });

// --- Licensing Commands ---

/**
 * Retrieves the current license status from the stored license file.
 * @returns A promise that resolves to the license object or null if not found.
 */
export const getLicenseStatus = () =>
    invoke<License | null>("get_license_status");

/**
 * Verifies a license key, and if valid, saves it to the config directory.
 * @param licenseKey The raw string content of the license file.
 * @returns A promise that resolves to the validated license object.
 */
export const verifyAndStoreLicense = (licenseKey: string) =>
    invoke<License>("verify_and_store_license", { licenseKey });

// --- System Commands ---

/**
 * A command that checks for the "APPIMAGE" environment variable on Linux
 * to determine the installation method.
 * @returns {Promise<string>} A promise that resolves to either "appimage" or "other".
 */
export function getLinuxInstallType(): Promise<string> {
    return invoke("get_linux_install_type");
}

/**
 * Checks the number of days the application has been in use.
 * @returns {Promise<number>} A promise that resolves to the number of days.
 */
export const getAppUsageDays = () => invoke<number>("get_app_usage_days");

// --- Template Commands ---

/**
 * Retrieves a list of all available templates.
 * @returns A promise that resolves to an array of PageHeader objects for the templates.
 */
export const listTemplates = () => invoke<PageHeader[]>("list_templates");

/**
 * Reads the raw content of a specific template file.
 * @param path The path of the template file to read.
 * @returns A promise that resolves to the string content of the template.
 */
export const readTemplate = (path: string) =>
    invoke<string>("read_template", { path });

/**
 * Saves content to a template file.
 * @param name The name of the template (without extension).
 * @param content The new content to save.
 * @returns A promise that resolves to the path of the saved template file.
 */
export const writeTemplate = (name: string, content: string) =>
    invoke<string>("write_template", { name, content });

/**
 * Deletes a template file.
 * @param path The path of the template file to delete.
 * @returns A promise that resolves when the template has been deleted.
 */
export const deleteTemplate = (path: string) =>
    invoke<void>("delete_template", { path });

// --- Custom Font Commands ---

/**
 * Scans the application's config directory for user-provided font files.
 * @returns A promise that resolves to an array of UserFont objects.
 */
export const getUserFonts = () => invoke<UserFont[]>("get_user_fonts");
